File: windows/src/Manifest.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace YzlabKurucu;

/// Sunucudaki kurulum manifesti. Model / token alan adi / Node adresi degisirse
/// SADECE sunucudaki JSON duzenlenir — yeni surum dagitmaya gerek kalmaz.
public sealed partial class Manifest
{
    public const string Url = "https://yapayzekalab.org/kurulum/codex.json";

    [JsonPropertyName("schemaVersion")] public int SchemaVersion { get; set; }
    [JsonPropertyName("api")] public ApiInfo Api { get; set; } = new();
    [JsonPropertyName("codex")] public CodexInfo Codex { get; set; } = new();
    [JsonPropertyName("node")] public NodeInfo Node { get; set; } = new();

    public sealed class ApiInfo
    {
        [JsonPropertyName("baseUrl")] public string BaseUrl { get; set; } = "";
        [JsonPropertyName("validateUrl")] public string ValidateUrl { get; set; } = "";
        [JsonPropertyName("authPrefix")] public string AuthPrefix { get; set; } = "Bearer ";
        [JsonPropertyName("keyPrefix")] public string KeyPrefix { get; set; } = "yzk_live_";
    }

    public sealed class ModelInfo
    {
        [JsonPropertyName("id")] public string Id { get; set; } = "";
        [JsonPropertyName("label")] public string Label { get; set; } = "";
        [JsonPropertyName("contextWindow")] public int ContextWindow { get; set; }
        public override string ToString() => Label;
    }

    public sealed class CodexInfo
    {
        [JsonPropertyName("npmPackage")] public string NpmPackage { get; set; } = "";
        [JsonPropertyName("catalogUrl")] public string CatalogUrl { get; set; } = "";
        [JsonPropertyName("catalogFile")] public string CatalogFile { get; set; } = "";
        [JsonPropertyName("profileName")] public string ProfileName { get; set; } = "";
        [JsonPropertyName("profileFile")] public string ProfileFile { get; set; } = "";
        [JsonPropertyName("defaultModel")] public string DefaultModel { get; set; } = "";
        [JsonPropertyName("models")] public List<ModelInfo> Models { get; set; } = new();
        [JsonPropertyName("profileTemplate")] public string ProfileTemplate { get; set; } = "";
        [JsonPropertyName("tokenField")] public string TokenField { get; set; } = "";
    }

    public sealed class NodePlatform
    {
        [JsonPropertyName("url")] public string Url { get; set; } = "";
        [JsonPropertyName("silentArgs")] public string SilentArgs { get; set; } = "";
    }

    public sealed class NodeInfo
    {
        [JsonPropertyName("windows")] public NodePlatform Windows { get; set; } = new();
    }

    /// Once sunucudan cek; ulasilamazsa gomulu surume dus (internet kesintisinde de kurulsun).
    public static async Task<(Manifest, bool canli)> LoadAsync()
    {
        try
        {
            using var http = new HttpClient { Timeout = TimeSpan.FromSeconds(10) };
            var json = await http.GetStringAsync(Url);
            var m = JsonSerializer.Deserialize<Manifest>(json);
            if (m is not null && m.Codex.Models.Count > 0) return (m, true);
        }
        catch { /* gomuluye dusulur */ }
        return (Embedded, false);
    }

    public static Manifest Embedded =>
        JsonSerializer.Deserialize<Manifest>(EmbeddedJson)
        ?? throw new InvalidOperationException("gomulu manifest bozuk — build hatasi");
}

File: macos/Sources/YzlabKurucu/Manifest.swift
import Foundation

/// Sunucudaki kurulum manifesti. Model / token alan adi / Node adresi degisirse
/// SADECE sunucudaki JSON duzenlenir — yeni surum dagitmaya gerek kalmaz.
/// `codex.catalogUrl` `{{CODEX_VERSION}}` tasiyabilir (Installer.katalogAdresi doldurur).
/// `schemaVersion`: canli manifest gomuluden ESKIYSE (yeni alanlar yok) gomulu kullanilir.
struct Manifest: Codable {
    struct Api: Codable {
        let baseUrl: String
        let validateUrl: String
        let authPrefix: String
        let keyPrefix: String
    }
    struct Model: Codable, Identifiable, Hashable {
        let id: String
        let label: String
        let contextWindow: Int
        private enum CodingKeys: String, CodingKey { case id, label, contextWindow }
    }
    struct Codex: Codable {
        let minVersion: String
        let npmPackage: String
        let catalogUrl: String
        let catalogFile: String
        let profileName: String
        let profileFile: String
        let launchCommand: String
        let defaultModel: String
        let models: [Model]
        let profileTemplate: String
        let tokenField: String
    }
    struct ClaudeModel: Codable, Identifiable, Hashable {
        let id: String
        let label: String
    }
    /// Claude Code (terminal + Claude masaustu uygulamasinin Code sekmesi): ikisi de
    /// ayni settings.json'i okur; profil mekanizmasi yok → yalniz `env` blogu duzenlenir.
    struct Claude: Codable {
        let npmPackage: String
        let configDir: String
        let settingsFile: String
        let baseUrl: String
        let defaultModel: String
        let smallFastModel: String
        let models: [ClaudeModel]
        let envTemplate: [String: String]
        let removeEnvKeys: [String]
        let launchCommand: String
    }
    struct NodePlatform: Codable {
        let url: String
        let silentArgs: String
    }
    struct Node: Codable {
        let macos: NodePlatform
    }

    let schemaVersion: Int
    let api: Api
    let codex: Codex
    let claude: Claude
    let node: Node

    static var manifestUrl: URL {
        #if DEBUG
        // Yalniz DEBUG derlemede: yerel test sunucusuna yonlendirme.
        // Dagitilan RELEASE ikilisinde bu dal DERLENMEZ — musterinin kurucusu
        // adresi asla disaridan degistirilemez.
        if let s = ProcessInfo.processInfo.environment["YZLAB_MANIFEST_URL"],
           let u = URL(string: s) { return u }
        #endif
        return URL(string: "https://yapayzekalab.org/kurulum/codex.json")!
    }

    /// Once sunucudan cek; ulasilamazsa / eski semaysa gomulu surume dus.
    /// Kurulumun internet kesintisinde de calismasi icin.
    static func load() async -> (Manifest, Bool) {
        var req = URLRequest(url: manifestUrl)
        req.timeoutInterval = 10
        req.cachePolicy = .reloadIgnoringLocalCacheData
        if let (data, resp) = try? await URLSession.shared.data(for: req),
           (resp as? HTTPURLResponse)?.statusCode == 200,
           let m = try? JSONDecoder().decode(Manifest.self, from: data),
           m.schemaVersion >= embedded.schemaVersion {
            return (m, true)
        }
        return (embedded, false)
    }

    static let embedded: Manifest = {
        guard let m = try? JSONDecoder().decode(Manifest.self, from: Data(embeddedJson.utf8)) else {
            fatalError("gomulu manifest bozuk — build hatasi")
        }
        return m
    }()
}
